e2e-topology-manager: single-NUMA-node test

ffromani opened the pull request:
**What type of PR is this?**
/kind cleanup

actually not sure about the proper type, please suggest

**What this PR does / why we need it**:
This is a followup of https://github.com/kubernetes/kubernetes/pull/86184 . I've been in touch with @vpickard in the past couple weeks, I'm working on adding more e2e tests for the topology manager
to fix the issue https://github.com/kubernetes/kubernetes/issues/83481

**Which issue(s) this PR fixes**:
Related to https://github.com/kubernetes/kubernetes/issues/83481

**Special notes for your reviewer**:
N/A

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

ffromani committed:
e2e: topomgr: explicit save the kubelet config

For the sake of readability, save the old Kubelet config
once.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: topomgr: add test infra

This patch all the testing infra and utilities needed
to run e2e topology manager tests. This include setup
a guaranteed pod which needs some devices.

The simplest real device available for the purpose
are the SRIOV devices, hence we use them.

This patch pulls the SRIOV device plugin from
the official, yet external, repository.
We do it as close as possible for the nvidia GPU plugin.

This patch also performs minor refactoring for some
test framework utilities, needed to support the new
e2e tests.

Finally, we add an empty e2e topology manager test,
to be completed by the next patch.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: topomgr: check pod resource alignment

This patch extends and completes the previously-added
empty topology manager test for single-NUMA node policy
by adding reporting in the test pod and checking
the resource alignment.

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
/sig testing

fromanirh commented:
I signed it

fromanirh commented:
/hold needs the final validation on my side

vpickard commented:
/cc @vpickard

ffromani committed:
e2e: topomgr: add more positive tests

this patch builds on the topology manager e2e infrastructure to
add more positive e2e test cases.

Signed-off-by: Francesco Romani <fromani@redhat.com>

vpickard commented:
/sig node

vpickard commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

vpickard commented:
@klueska @nolancon

vpickard commented:
/cc @derekwaynecarr Would it be possible to get an ok-to-test on this one?

nolancon commented:
/ok-to-test

vpickard commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

fromanirh commented:
hi, thanks for the ok-to-test. I'm aware of failures - all related to the build step AFAICT- and I'll work to fix them ASAP.

fromanirh commented:
I've force-pushed a new patchset to squash in the bazel build fixes. I'll try hard not to force-push again.

fromanirh commented:
now `make bazel-build` completes correctly on my test box, let's see if the bots like it.

fromanirh commented:
/hold
still need the last round of validation on my side (final testing on multi node NUMA)

fromanirh commented:
/retest

ffromani committed:
e2e: topomgr: autodetect SRIOV resource to use

The SRIOV device plugin can create different resources depending
on both the hardware present on the system and the configuration.
As long as we have at least one SRIOV device, the tests don't actually
care about which specific device is.

Previously, the test hardcoded the most common intel SRIOV device
identifier. This patch lifts the restriction and let the test
autodetect and use what's available.

Signed-off-by: Francesco Romani <fromani@redhat.com>

mattjmcnaughton reviewed (commented):
I don't have a ton of context on the NUMA stuff, but am happy to help with resolving the `kubernetes-verify` issues if you get stuck at all - just let me know :)

Thanks for the additional test coverage.

fromanirh commented:
> I don't have a ton of context on the NUMA stuff, but am happy to help with resolving the `kubernetes-verify` issues if you get stuck at all - just let me know :)
> 
> Thanks for the additional test coverage.

Thanks! So far the failed `kubernetes-verify` failures are pretty self-explanatory. I guess I just need to get in the habit to run the linter stage on my box before to submit :)

fromanirh commented:
patches fixing `kubernetes-verify` test failures intentionally unsquashed to avoid force-pushes. I'll happily squash them in once the reviewers request it.

fromanirh commented:
/hold cancel

fromanirh commented:
/retest

fromanirh commented:
/test pull-kubernetes-e2e-gce-100-performance

fromanirh commented:
/retest

fromanirh commented:
/retest

ffromani committed:
e2e: topomgr: add option to specify the SRIOV conf

We cannot anticipate all the possible configurations
needed by the SRIOV device plugin: there is too much variety.

Hence, we need to allow the test environment to supply
a host-specific ConfigMap to properly configure the device
plugin and avoid false negatives.

We still provide a the default config map as fallback and reference.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: topomgr: initial negative tests

Negative tests is when we request a gu Pod we know the system cannot
fullfill - hence we expect rejection from the topology manager.

Unfortunately, besides the trivial case of excessive cores (request
more socket than a NUMA node provides) we cannot easily test the
devices, because crafting a proper pod will require detailed knowledge
of the hw topology.

Let's consider a hypotetical two-node NUMA system with two PCIe busses,
one per NUMA node, with a SRIOV device on each bus.
A proper negative test would require two SRIOV device, that the system
can provide but not on the same single NUMA node.
Requiring for example three devices (one more than the system provides)
will lead to a different, legitimate admission error.

For these reasons we bootstrap the testing infra for the negative tests,
but we add just the simplest one.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: topomgr: get and use topology hints from conf

TO properly implement some e2e tests, we need to know
some basic topology facts about the system running the tests.
The bare minimum we need to know is how many PCI SRIOV devices
are attached to which NUMA node.

This way we know which core we can reserve for kube services,
and which NUMA socket we can take to test full socket reservation.

To let the tests know the PCI device topology, we use annotations
in the SRIOV device plugin ConfigMap we need anyway.
The format is

```yaml
  metadata:
    annotations:
      pcidevice_node0: "2"
      pcidevice_node1: "0"
```

with one annotation per NUMA node in the system.

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
Hi, I have more tests in the pipeline. Since no reviews happened so far, I'm going to add the patches to this PR, so reviewers can have the complete picture of everything I want to add. I'm going to upload the patches.

fromanirh commented:
> @fromanirh: PR needs rebase.

Uhm, since I need a rebase I'll squash the fixes in.

fromanirh commented:
fixes squashed in, changes rebased and patch queue empty. Please review!

fromanirh commented:
bad rebase, fixing

fromanirh commented:
/remove-label  sig/scheduling

fromanirh commented:
/remove-area sig/scheduling

fromanirh commented:
/remove sig/scheduling sig/cluster-lifecycle

fromanirh commented:
/remove-sig scheduling cluster-lifecycle

fromanirh commented:
/assign @klueska @derekwaynecarr

fromanirh commented:
last push includes the golint changes that should make the `verify` lane pass.

fromanirh commented:
The PR needs some explicit configuration (due to the complexity of the possible HW setups), documentation here: https://github.com/fromanirh/community/commit/1c7b43dd0fff0d0f1aad18a6a4ac7ead33716f39 - will file a PR asap

fromanirh commented:
/test pull-kubernetes-e2e-gce

vpickard commented on test/e2e_node/topology_manager_test.go:
```diff
@@ -350,27 +499,231 @@ func runTopologyManagerSuiteTests(f *framework.Framework) {
+		framework.Logf("Node status: %v", node.Status.Capacity)
+		sriovResourceName, sriovResourceAmount = findSRIOVResource(node)
+		return sriovResourceAmount > 0
+	}, 5*time.Minute, framework.Poll).Should(gomega.BeTrue())
```
From our slack chat, consider changing this timeout from 5 mins to 2 mins.

fromanirh replied:
Done

vpickard commented on test/e2e_node/topology_manager_test.go:
```diff
@@ -350,27 +499,231 @@ func runTopologyManagerSuiteTests(f *framework.Framework) {
+
+	sriovResourceName := ""
+	var sriovResourceAmount int64
+	ginkgo.By("Waiting for devices to become available on the local node")
```
It may be a good idea to document, perhaps with a comment here, and in test doc code, that VFs should already be configure on the system under test. On my setup, I did not have any VFs configured (wiped out from doing other stuff), and the test spun here for 5 mins. 

Perhaps some example config/commands would be helpful. For example:

cat /sys/devices/pci0000:00/0000:00:01.0/0000:01:00.1/sriov_numvfs
echo 7 > /sys/devices/pci0000:00/0000:00:01.0/0000:01:00.1/sriov_numvfs

fromanirh replied:
Will post a PR against https://github.com/kubernetes/community/blob/master/contributors/devel/sig-node/e2e-node-tests.md - but will also add a comment here.

vpickard commented on test/e2e_node/topology_manager_test.go:
```diff
@@ -122,9 +201,79 @@ func configureTopologyManagerInKubelet(f *framework.Framework, policy string) {
+}
+
+func findSRIOVResource(node *v1.Node) (string, int64) {
+	re := regexp.MustCompile(`^intel.com/.*sriov.*`)
```
SR-IOV device plugin also supports Mellanox devices. I think it would be a good idea if this could be expanded to work with either Intel or Mellanox.

fromanirh replied:
Yup, already improved here: https://github.com/kubernetes/kubernetes/pull/87645/commits/0fad14ffe1c3daadf2a6d47c3ed68e8fd8e80bfb

vpickard commented on test/e2e_node/topology_manager_test.go:
```diff
@@ -350,27 +499,231 @@ func runTopologyManagerSuiteTests(f *framework.Framework) {
+func runTopologyManagerNodeAlignmentSuiteTests(f *framework.Framework, configMap *v1.ConfigMap, reservedSystemCPUs string, numaNodes, coreCount int) {
+	var err error
+
+	ginkgo.By(fmt.Sprintf("Creating configMap %v/%v", metav1.NamespaceSystem, configMap.Name))
```
Perhaps consider moving the SRIOV setup to separate function (setupSRIOVConf()... or something like that). 

Eventually, we will have support for other devices (memory manager, GPU). It seems a little cleaner to have the device setup separately from the actual tests.

fromanirh replied:
Done

ffromani committed:
e2r: topomgr: improve the test logs

Add clarification to which test is doing what, to make
the test output easier to understand.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: topomgr: reduce node readiness timeout

Five minutes was initially used only to be overcautious.
From my experiments, the node is ready in usually less than a minute.
Double it to give some buffer space.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: topomgr: better check for AffinityError

Add a helper function to check if a Pod failed
admission for Topology Affinity Error.
So far we only check the Status.Reason.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: topomgr: use constants for test limits

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented on test/e2e_node/topology_manager_test.go:
```diff
@@ -350,27 +463,128 @@ func runTopologyManagerSuiteTests(f *framework.Framework) {
+		coreCount := detectCoresPerSocket()
+		sriovdevCount := detectSRIOVDevices()
+
+		if numaNodes < 2 {
```
Done

ffromani committed:
e2e: topomgr: introduce sriov setup/teardown funcs

Reorganize the code with setup and teardown functions,
to make room for the future addition of more device plugin
support, and to make the code a bit tidier.

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
/retest

derekwaynecarr commented:
Can you document the setup required for this test?

Do we need the actual SR-IOV device plugin or can we have something that is mocked?

I will let @vpickard finish his review and provide and LGTM before doing a final pass for approval.  Thanks for your contribution.

vpickard commented on test/e2e_node/topology_manager_test.go:
```diff
@@ -350,27 +501,252 @@ func runTopologyManagerSuiteTests(f *framework.Framework) {
+}
+
+func isTopologyAffinityError(pod *v1.Pod) bool {
+	re := regexp.MustCompile(`Topology.*Affinity.*Error`)
```
The negative test was failing in my setup. Debugged it, and could see from older version where it was failing. Code was checking for Topology Affinity Error, from below const definition. I see you have fixed this in latest update. Nice!

topologyError = "Topology Affinity Error" // XXX do we have a proper constant?

fromanirh replied:
Sorry for causing you troubles. I'm not very happy with the current test but it's the best I could think of, lacking constants or symbols I could depend on :\

nolancon replied:
Not sure if this is relevant, but just FYI: Changes introduced in https://github.com/kubernetes/kubernetes/pull/87758 mean the error is now "TopologyAffinityError" as opposed to previous "Topology Affinity Error".

fromanirh replied:
Ack, I missed this, thanks for pointing it out. This is the reason why it worked before and then failed. Anyway I'd keep the regexp match - lacking better alternatives - because its' a little less fragile than the string matching I had before.

vpickard commented on test/e2e_node/topology_manager_test.go:
```diff
@@ -350,27 +501,252 @@ func runTopologyManagerSuiteTests(f *framework.Framework) {
+	framework.Logf("deleting the SRIOV device plugin pod %s/%s and waiting for container %s removal",
+		dpPod.Namespace, dpPod.Name, dpPod.Spec.Containers[0].Name)
+	deletePodInNamespace(f, dpPod.Namespace, dpPod.Name)
+	waitForContainerRemoval(dpPod.Spec.Containers[0].Name, dpPod.Name, dpPod.Namespace)
```
I'm seeing this failure in my setup when I run the test. Looks like 2 min timeout waiting to remove container...

Feb 10 23:39:53.117: INFO: Pod "gu-pod" satisfied condition "Failed"
Feb 10 23:39:53.125: INFO: Waiting for pod gu-pod to disappear
Feb 10 23:39:53.127: INFO: Pod gu-pod no longer exists
Feb 10 23:39:53.127: INFO: deleting the SRIOV device plugin pod kube-system/sriov-device-plugin and waiting for container kube-sriovdp removal

• Failure [498.096 seconds]
[sig-node] Topology Manager [Serial] [Feature:TopologyManager][NodeAlphaFeature:TopologyManager]
_output/local/go/src/k8s.io/kubernetes/test/e2e_node/framework.go:23
  With kubeconfig updated to static CPU Manager policy run the Topology Manager tests
  _output/local/go/src/k8s.io/kubernetes/test/e2e_node/topology_manager_test.go:739
    run Topology Manager node alignment test suite [It]
    _output/local/go/src/k8s.io/kubernetes/test/e2e_node/topology_manager_test.go:695

    Timed out after 120.000s.
    Expected                                                                                                                                                            
        <bool>: false                                                                                                                                                   
    to be true

fromanirh replied:
> Can you document the setup required for this test?
> 
> Do we need the actual SR-IOV device plugin or can we have something that is mocked?
> 
> I will let @vpickard finish his review and provide and LGTM before doing a final pass for approval. Thanks for your contribution.

Thanks for your comments! I added the requirements and some documentation here: https://github.com/kubernetes/community/pull/4503

Is there a better place?

fromanirh replied:
We investigated on Victor's testing machine, and it seems a local glitch. We'll check again in the near future.

vpickard replied:
Yes, it was a docker issue on my setup. I have it working now.

fromanirh commented:
/retest

fromanirh commented:
/retest

fromanirh commented:
/test pull-kubernetes-e2e-gce

fromanirh commented:
> Can you document the setup required for this test?
> 
> Do we need the actual SR-IOV device plugin or can we have something that is mocked?
> 
> I will let @vpickard finish his review and provide and LGTM before doing a final pass for approval. Thanks for your contribution.

Thanks for your comments! I added the requirements and some documentation here: kubernetes/community#4503

Is there a better place?

vpickard commented:
All review comments have been addressed. Testing passes on my local setup, as well as in GCE.

/lgtm

vpickard commented:
> Can you document the setup required for this test?
> 
> Do we need the actual SR-IOV device plugin or can we have something that is mocked?
> 
> I will let @vpickard finish his review and provide and LGTM before doing a final pass for approval. Thanks for your contribution.

@derekwaynecarr I have finished my review, LGTM. Pleae review when ready.

fejta commented:
/approve

nolancon commented:
/lgtm
